// @sudoku/stores/session.js
// Store 适配层：将领域对象桥接到 Svelte 响应式系统
// 闭包中持有 Game 实例，对外暴露 writable store 快照
import { decodeSencode } from '@sudoku/sencode';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

function createEmptyGrid() {
	return Array.from({ length: 9 }, () => Array(9).fill(0));
}

// 从 Game 实例生成 plain object 快照，供 Svelte store 消费
function createSnapshot(game) {
	return {
		baseGrid:     game.getInitialGrid(),
		userGrid:     game.getGrid(),
		invalidCells: game.getInvalidCells().map(({ row, col }) => `${col},${row}`),
		canUndo:      game.canUndo(),
		canRedo:      game.canRedo(),
		won:          game.isSolved(),
	};
}

function createGameSession() {
	const session = writable(null);
	let game = createGame({
		sudoku: createSudoku(createEmptyGrid()),
	});

	// 每次领域变更后调用，生成新快照推送到 store 触发 UI 更新
	function publish() {
		session.set(createSnapshot(game));
	}

	function loadPuzzle(puzzleGrid) {
		game = createGame({
			sudoku: createSudoku(puzzleGrid),
		});
		publish();
		return true;
	}

	function guess(move) {
		const changed = game.guess(move);

		if (changed) {
			publish();
		}

		return changed;
	}

	function undo() {
		const changed = game.undo();

		if (changed) {
			publish();
		}

		return changed;
	}

	function redo() {
		const changed = game.redo();

		if (changed) {
			publish();
		}

		return changed;
	}

	publish();

	return {
		subscribe: session.subscribe,

		loadPuzzle,

		generate(difficulty) {
			return loadPuzzle(generateSudoku(difficulty));
		},

		decodeSencode(sencode) {
			return loadPuzzle(decodeSencode(sencode));
		},

		guess,

		applyHint(position) {
			try {
				const solvedSudoku = solveSudoku(game.getGrid());
				return guess({
					row:   position.y,
					col:   position.x,
					value: solvedSudoku[position.y][position.x],
				});
			} catch {
				return false;
			}
		},

		undo,

		redo,
	};
}

export const gameSession = createGameSession();
